File: leetcode/0812_largest_triangle_area.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <cmath>
#include <vector>

using namespace std;

class Solution {
 public:
    double largestTriangleArea(vector<vector<int>>& points) {
        double result = 0;

        for (int i = 0; i < points.size() - 2; i++) {
            for (int j = i + 1; j < points.size() - 1; j++) {
                for (int k = j + 1; k < points.size(); k++) {
                    double x_a = points[i][0], y_a = points[i][1];
                    double x_b = points[j][0], y_b = points[j][1];
                    double x_c = points[k][0], y_c = points[k][1];
                    double area = abs((x_a - x_c) * (y_b - y_a) - (x_a - x_b) * (y_c - y_a)) / 2;
                    if (area > result) {
                        result = area;
                    }
                }
            }
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    vector<vector<int>> points = {{0, 0}, {0, 1}, {1, 0}, {0, 2}, {2, 0}};

    assert(2 == solution.largestTriangleArea(points));
}

File: leetcode/0812_largest_triangle_area.py
# -*- coding: utf-8 -*-

import itertools


class Solution:
    def largestTriangleArea(self, points):
        result = 0

        for (x_a, y_a), (x_b, y_b), (x_c, y_c) in itertools.combinations(points, 3):
            area = abs((x_a - x_c) * (y_b - y_a) - (x_a - x_b) * (y_c - y_a)) / 2
            if area > result:
                result = area

        return result


if __name__ == '__main__':
    solution = Solution()

    assert 2 == solution.largestTriangleArea([[0, 0], [0, 1], [1, 0], [0, 2], [2, 0]])
    assert 0.5 == solution.largestTriangleArea([[1, 0], [0, 0], [0, 1]])
